File: Examples/kotlin/basics.kt
data class Point(val x: Int, val y: Int)

enum class Color { RED, GREEN, BLUE }

fun fib(n: Int): Int = if (n < 2) n else fib(n - 1) + fib(n - 2)

fun describe(n: Int): String = when {
    n < 0 -> "negative"
    n == 0 -> "zero"
    n < 10 -> "small"
    else -> "large"
}

class Counter(private var value: Int = 0) {
    fun increment() { value++ }
    fun get(): Int = value
}

fun main() {
    println("Hello, Kotlin!")
    var sum = 0
    for (i in 1..10) sum += i
    println("sum = $sum")
    println(fib(15))

    val xs = listOf(5, 2, 9, 1)
    println(xs.sorted())
    println(xs.map { it * it })
    println(xs.filter { it > 2 })
    println(xs.sum())
    println(xs.maxOrNull())

    val p = Point(3, 4)
    println(p)
    println(p.x + p.y)

    for (c in listOf(Color.RED, Color.GREEN)) println(c)

    println(describe(-1))
    println(describe(0))
    println(describe(5))
    println(describe(50))

    val c = Counter(10)
    c.increment()
    c.increment()
    println(c.get())

    val m = mutableMapOf("a" to 1, "b" to 2)
    m["c"] = 3
    println(m["a"]!! + m["c"]!!)
    println(m.keys.sorted())

    val words = listOf("pear", "apple", "fig")
    println(words.sortedBy { it.length })
    println(words.joinToString(", "))

    val s = "Hello, World"
    println(s.length)
    println(s.uppercase())
    println(s.substring(7))
    println(s.split(", "))

    var i = 0
    while (i < 3) { print("$i "); i++ }
    println()
    println(listOf(1,2,3).fold(0) { acc, v -> acc + v })
}

File: Examples/cpp/basics.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <algorithm>
using namespace std;

struct Point {
    int x;
    int y;
};

class Counter {
private:
    int value;
public:
    Counter(int start) { value = start; }
    void increment() { value++; }
    int get() const { return value; }
};

int fib(int n) {
    if (n < 2) return n;
    return fib(n - 1) + fib(n - 2);
}

int main() {
    cout << "Hello, C++!" << endl;

    int sum = 0;
    for (int i = 1; i <= 10; i++) sum += i;
    cout << "sum = " << sum << endl;
    cout << fib(15) << endl;

    vector<int> v;
    for (int i = 5; i > 0; i--) v.push_back(i * i);
    sort(v.begin(), v.end());
    for (int x : v) cout << x << " ";
    cout << endl;

    string s = "hello world";
    cout << s.size() << " " << s.substr(6) << endl;

    map<string, int> counts;
    counts["a"] = 1;
    counts["b"] = 2;
    cout << counts["a"] + counts["b"] << endl;

    Counter c(10);
    c.increment();
    c.increment();
    cout << c.get() << endl;

    double d = 1.0 / 3.0;
    cout << d << endl;
    cout << 7 / 2 << " " << 7 % 3 << endl;

    int arr[5] = {3, 1, 4, 1, 5};
    int total = 0;
    for (int i = 0; i < 5; i++) total += arr[i];
    cout << total << endl;

    auto square = [](int n) { return n * n; };
    cout << square(7) << endl;

    printf("%d %s %.2f\n", 42, "ok", 3.14159);
    return 0;
}
